csharp/src/Ice/ConnectionFactory.cs

```diff
@@ -71,208 +71,192 @@ public void Remove(Connection connection)
+            IReadOnlyList<Connector> excludedConnectors,
+            CancellationToken cancel)
+        {
+            // Purge expired transport failures
+            DateTime expirationDate = DateTime.Now - TimeSpan.FromSeconds(5);
+            foreach ((Connector connector, DateTime date) in _transportFailures)
+            {
+                if (date <= expirationDate)
+                {
+                    _ = _transportFailures.TryRemove(connector, out DateTime _);
+                }
+            }
+
+            // For each endpoint, obtain the set of connectors. This might block if DNS lookups are required to resolve
+            // an endpoint hostname into connector addresses.
+            var connectors = new List<Connector>();
             foreach (Endpoint endpoint in endpoints)
             {
                 try
                 {
-                    IEnumerable<IConnector> endpointConnectors = await endpoint.ConnectorsAsync(cancel).ConfigureAwait(false);
-                    foreach (IConnector connector in endpointConnectors)
+                    // TODO should we return a single connector from ConnectorsAsync corresponding with the first
+                    // DNS address?
+                    IEnumerable<Connector> endpointConnectors =
+                        await endpoint.ConnectorsAsync(cancel).ConfigureAwait(false);
+                    foreach (Connector connector in endpointConnectors)
                     {
-                        connectors.Add((connector, endpoint));
+                        if (!excludedConnectors.Contains(connector))
+                        {
+                            connectors.Add(connector);
+                        }
                     }
                 }
                 catch (CommunicatorDisposedException)
                 {
                     throw; // No need to continue
                 }
-                catch (Exception ex)
+                catch (Exception)
                 {
-                    bool last = endpoint == endpoints[endpoints.Count - 1];
-
-                    TraceLevels traceLevels = _communicator.TraceLevels;
-                    if (traceLevels.Transport >= 2)
-                    {
-                        _communicator.Logger.Trace(TraceLevels.TransportCategory, last ?
-                            $"couldn't resolve endpoint host and no more endpoints to try\n{ex}" :
-                            $"couldn't resolve endpoint host, trying next endpoint\n{ex}");
-                    }
-
-                    if (connectors.Count == 0 && last)
-                    {
-                        // If this was the last endpoint and we didn't manage to get a single connector, we're done.
-                        throw;
-                    }
                 }
             }
 
+            // If all the connectors are in the excludedConnectors throw NoEndpointException to indicate the caller
+            // there is no more endpoints to try.
```

Comment:
Right maybe we should make NoEndpointException a TransportException, and set its retryable policy to OtherReplica when using cached endpoints

Reply:
Hmm shouldn't we just let go through the DNS failure (the last one for example) to actually let the user know the reason of the connection failure on the endpoints?

Reply:
Fixed the code to throw the last exception in case there are no connectors, NoEndpointException will be only raised when all connectors are excluded.